Source/WebCore/platform/audio/SharedAudioDestination.cpp

```diff
@@ -0,0 +1,315 @@
+#include <wtf/WTFSemaphore.h>
+#include <wtf/WeakPtr.h>
+#include <wtf/WorkQueue.h>
+
+namespace WebCore {
+
+class SharedAudioDestinationAdapter : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<SharedAudioDestinationAdapter>, public AudioIOCallback {
+public:
+    using AudioDestinationCreationFunction = SharedAudioDestination::AudioDestinationCreationFunction;
+    static Ref<SharedAudioDestinationAdapter> ensureAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&& ensureFunction);
+    ~SharedAudioDestinationAdapter();
+
+    void addRenderer(SharedAudioDestination&, CompletionHandler<void(bool)>&&);
+    void removeRenderer(SharedAudioDestination&, CompletionHandler<void(bool)>&&);
+
+    unsigned framesPerBuffer() const
+    {
+        return m_workBus->length();
+    }
+
+private:
+    using AdapterKey = std::tuple<unsigned, float>;
+    using AdapterMap = HashMap<AdapterKey, ThreadSafeWeakPtr<SharedAudioDestinationAdapter>>;
+    static AdapterMap& sharedMap();
+
+    SharedAudioDestinationAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&&);
+
+    void render(AudioBus* sourceBus, AudioBus* destinationBus, size_t framesToProcess, const AudioIOPosition& outputPosition) final;
+    void isPlayingDidChange() final { }
+
+    void configureRenderThread(CompletionHandler<void(bool)>&&);
+
+    unsigned m_numberOfOutputChannels;
+    float m_sampleRate;
+
+    Ref<AudioDestination> m_destination;
+    Ref<AudioBus> m_workBus;
+    Ref<WorkQueue> m_configurationQueue;
+
+    bool m_started { false };
+
+    Lock m_renderLock;
+    Semaphore m_configurationSemaphore;
+
+    // Only accessed on m_configurationQueue:
+    Vector<RefPtr<SharedAudioDestination>> m_renderers;
+
+    bool m_needsConfiguration WTF_GUARDED_BY_LOCK(m_renderLock) { true };
+    Vector<RefPtr<SharedAudioDestination>> m_newRenderers WTF_GUARDED_BY_LOCK(m_renderLock);
+
+    // Only accessed on the audio thread:
+    Vector<RefPtr<SharedAudioDestination>> m_configuredRenderers;
+};
+
+auto SharedAudioDestinationAdapter::sharedMap() -> AdapterMap&
+{
+    static MainThreadNeverDestroyed<AdapterMap> map;
+    return map;
+}
+
+Ref<SharedAudioDestinationAdapter> SharedAudioDestinationAdapter::ensureAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&& ensureFunction)
+{
+    auto key = std::make_tuple(numberOfOutputChannels, sampleRate);
+    auto results = sharedMap().find(key);
+    if (results != sharedMap().end()) {
+        if (auto existingAdapter = results->value.get())
+            return existingAdapter.releaseNonNull();
+    }
+
+    Ref<SharedAudioDestinationAdapter> newAdapter = adoptRef(*new SharedAudioDestinationAdapter(numberOfOutputChannels, sampleRate, WTFMove(ensureFunction)));
+    auto weakAdapter = ThreadSafeWeakPtr<SharedAudioDestinationAdapter> { newAdapter.get() };
+    sharedMap().set(key, WTFMove(weakAdapter));
```

Comment:
Wouldn't this work?
```
sharedMap().set(key, newAdapter.get());
```
Why do we need to constructor the `ThreadSafeWeakPtr` explicitly. And if we really have to, maybe this would work?
```
ThreadSafeWeakPtr weakAdapter { newAdapter.get() };
```

Reply:
This code was super fiddly, and automatic type deduction was breaking left and right when I wrote it. I originally had this: 

```
sharedMap().set(key, newAdapter.get());
```

And it lead to a compile error deducing the type of the value parameter; hence the rather explicit line 103.